File: pyscript.core/src/stdlib/pyscript/storage.py
from polyscript import storage as _storage
from pyscript.flatted import parse as _parse, stringify as _stringify


# convert a Python value into an IndexedDB compatible entry
def _to_idb(value):
    if value is None:
        return _stringify(["null", 0])
    if isinstance(value, (bool, float, int, str, list, dict, tuple)):
        return _stringify(["generic", value])
    if isinstance(value, bytearray):
        return _stringify(["bytearray", [v for v in value]])
    if isinstance(value, memoryview):
        return _stringify(["memoryview", [v for v in value]])
    raise f"Unexpected value: {value}"


# convert an IndexedDB compatible entry into a Python value
def _from_idb(value):
    (
        kind,
        result,
    ) = _parse(value)
    if kind == "null":
        return None
    if kind == "generic":
        return result
    if kind == "bytearray":
        return bytearray(result)
    if kind == "memoryview":
        return memoryview(bytearray(result))
    return value


async def storage(*args):
    if len(args):
        (name,) = args
    else:
        name = "core"

    store = await _storage(f"@pyscript/{name}")
    known = {k: _from_idb(v) for k, v in store.entries()}

    class Storage(dict):
        def __init__(self, known):
            super().__init__(known)

        def __delitem__(self, attr):
            store.delete(attr)
            super().__delitem__(attr)

        def __setitem__(self, attr, value):
            store.set(attr, _to_idb(value))
            super().__setitem__(attr, value)

        def clear(self):
            store.clear()
            super().clear()

        async def sync(self):
            await store.sync()

    return Storage(known)

File: core/src/storage.js
import IDBMapSync from "@webreflection/idb-map/sync";
import { parse, stringify } from "flatted";

const { isView } = ArrayBuffer;

const to_idb = (value) => {
    if (value == null) return stringify(["null", 0]);
    /* eslint-disable no-fallthrough */
    switch (typeof value) {
        case "object": {
            if (isView(value)) return stringify(["memoryview", [...value]]);
            if (value instanceof ArrayBuffer)
                return stringify(["bytearray", [...new Uint8Array(value)]]);
        }
        case "string":
        case "number":
        case "boolean":
            return stringify(["generic", value]);
        default:
            throw new TypeError(`Unexpected value: ${String(value)}`);
    }
};

const from_idb = (value) => {
    const [kind, result] = parse(value);
    if (kind === "null") return null;
    if (kind === "generic") return result;
    if (kind === "bytearray") return new Uint8Array(value).buffer;
    if (kind === "memoryview") return new Uint8Array(value);
    return value;
};

// this export simulate pyscript.storage exposed in the Python world
export const storage = async (name) => {
    if (!name) throw new SyntaxError("The storage name must be defined");

    const store = new IDBMapSync(`@pyscript/${name}`);
    const map = new Map();
    await store.sync();
    for (const [k, v] of store.entries()) map.set(k, from_idb(v));

    const clear = () => {
        map.clear();
        store.clear();
    };

    const sync = async () => {
        await store.sync();
    };

    return new Proxy(map, {
        ownKeys: (map) => [...map.keys()],
        has: (map, name) => map.has(name),
        get: (map, name) => {
            if (name === "clear") return clear;
            if (name === "sync") return sync;
            return map.get(name);
        },
        set: (map, name, value) => {
            map.set(name, value);
            store.set(name, to_idb(value));
            return true;
        },
        deleteProperty: (map, name) => {
            if (map.has(name)) {
                map.delete(name);
                store.delete(name);
            }
            return true;
        },
    });
};
